[YAML-CLI2] polars.exceptions.ComputeError: datatypes of join keys don't match

### Background / motivation

(In yaml-cli2 branch.)

Calina and I each independently encountered this error, after making trivial changes to the code.

Short version of error:
```
FAILED pdr_backend/lake/test/test_ohlcv_data_factory.py::test_get_mergedohlcv_df_happypath - polars.exceptions.ComputeError: datatypes of join keys don't match - `timestamp`: null on left does not match `timestamp`: i64 on right
```

Full traceback: see below.

Calina mentioned that it fails sporadically for her (a recent occurrence, in the last two weeks). I just got back from holidays; I've run CI twice; this test failed once. So for me failures are sporadic too (1/2 times).

Calina mentioned that it might be because sometimes Kraken returns good data from ccxt, and other times not. And in my experience, Kraken usually (always?) gives bad data. Therefore as part of the fix, I'll make sure we're testing for getting good data from ccxt kraken, and bad data. (Assisted by mocks.)

### Full traceback

From [this log](https://github.com/oceanprotocol/pdr-backend/actions/runs/7472573608/job/20335037499):

(The log has many async-related warnings too. They're omitted from here; [#507](https://github.com/oceanprotocol/pdr-backend/issues/507) is handling them.)

```text
=================================== FAILURES ===================================
______________________ test_get_mergedohlcv_df_happypath _______________________

tmpdir = local('/tmp/pytest-of-runner/pytest-0/test_get_mergedohlcv_df_happyp0')

    @enforce_types
    def test_get_mergedohlcv_df_happypath(tmpdir):
        """Is get_mergedohlcv_df() executing e2e correctly? Incl. call to exchange.
    
        It may fail if the exchange is temporarily misbehaving, which
          shows up as a FileNotFoundError.
        So give it a few tries if needed.
        """
        n_tries = 5
        for try_i in range(n_tries - 1):
            try:
>               _test_get_mergedohlcv_df_happypath(tmpdir)

pdr_backend/lake/test/test_ohlcv_data_factory.py:158: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/enforce_typing/decorator.py:29: in wrapper
    return func(*args, **kwargs)
pdr_backend/lake/test/test_ohlcv_data_factory.py:182: in _test_get_mergedohlcv_df_happypath
    mergedohlcv_df = factory.get_mergedohlcv_df()
pdr_backend/lake/ohlcv_data_factory.py:84: in get_mergedohlcv_df
    mergedohlcv_df = merge_rawohlcv_dfs(rawohlcv_dfs)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/enforce_typing/decorator.py:29: in wrapper
    return func(*args, **kwargs)
pdr_backend/lake/merge_df.py:32: in merge_rawohlcv_dfs
    merged_df = _add_df_col(merged_df, merged_col, raw_df, raw_col)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/enforce_typing/decorator.py:29: in wrapper
    return func(*args, **kwargs)
pdr_backend/lake/merge_df.py:58: in _add_df_col
    merged_df = merged_df.join(newraw_df, on="timestamp", how="outer")
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/polars/dataframe/frame.py:6339: in join
    self.lazy()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <LazyFrame [5 cols, {"timestamp": Null … "kraken:BTC/USDT:low": Float64}] at 0x7F67ECA6D670>

    def collect(
        self,
        *,
        type_coercion: bool = True,
        predicate_pushdown: bool = True,
        projection_pushdown: bool = True,
        simplify_expression: bool = True,
        slice_pushdown: bool = True,
        comm_subplan_elim: bool = True,
        comm_subexpr_elim: bool = True,
        no_optimization: bool = False,
        streaming: bool = False,
        background: bool = False,
        _eager: bool = False,
    ) -> DataFrame | InProcessQuery:
        """
        Materialize this LazyFrame into a DataFrame.
    
        By default, all query optimizations are enabled. Individual optimizations may
        be disabled by setting the corresponding parameter to `False`.
    
        Parameters
        ----------
        type_coercion
            Do type coercion optimization.
        predicate_pushdown
            Do predicate pushdown optimization.
        projection_pushdown
            Do projection pushdown optimization.
        simplify_expression
            Run simplify expressions optimization.
        slice_pushdown
            Slice pushdown optimization.
        comm_subplan_elim
            Will try to cache branching subplans that occur on self-joins or unions.
        comm_subexpr_elim
            Common subexpressions will be cached and reused.
        no_optimization
            Turn off (certain) optimizations.
        streaming
            Process the query in batches to handle larger-than-memory data.
            If set to `False` (default), the entire query is processed in a single
            batch.
    
            .. warning::
                This functionality is currently in an alpha state.
    
            .. note::
                Use :func:`explain` to see if Polars can process the query in streaming
                mode.
        background
            Run the query in the background and get a handle to the query.
            This handle can be used to fetch the result or cancel the query.
    
        Returns
        -------
        DataFrame
    
        See Also
        --------
        fetch: Run the query on the first `n` rows only for debugging purposes.
        explain : Print the query plan that is evaluated with collect.
        profile : Collect the LazyFrame and time each node in the computation graph.
        polars.collect_all : Collect multiple LazyFrames at the same time.
        polars.Config.set_streaming_chunk_size : Set the size of streaming batches.
    
        Examples
        --------
        >>> lf = pl.LazyFrame(
        ...     {
        ...         "a": ["a", "b", "a", "b", "b", "c"],
        ...         "b": [1, 2, 3, 4, 5, 6],
        ...         "c": [6, 5, 4, 3, 2, 1],
        ...     }
        ... )
        >>> lf.group_by("a").agg(pl.all().sum()).collect()  # doctest: +SKIP
        shape: (3, 3)
        ┌─────┬─────┬─────┐
        │ a   ┆ b   ┆ c   │
        │ --- ┆ --- ┆ --- │
        │ str ┆ i64 ┆ i64 │
        ╞═════╪═════╪═════╡
        │ a   ┆ 4   ┆ 10  │
        │ b   ┆ 11  ┆ 10  │
        │ c   ┆ 6   ┆ 1   │
        └─────┴─────┴─────┘
    
        Collect in streaming mode
    
        >>> lf.group_by("a").agg(pl.all().sum()).collect(
        ...     streaming=True
        ... )  # doctest: +SKIP
        shape: (3, 3)
        ┌─────┬─────┬─────┐
        │ a   ┆ b   ┆ c   │
        │ --- ┆ --- ┆ --- │
        │ str ┆ i64 ┆ i64 │
        ╞═════╪═════╪═════╡
        │ a   ┆ 4   ┆ 10  │
        │ b   ┆ 11  ┆ 10  │
        │ c   ┆ 6   ┆ 1   │
        └─────┴─────┴─────┘
    
        """
        if no_optimization or _eager:
            predicate_pushdown = False
            projection_pushdown = False
            slice_pushdown = False
            comm_subplan_elim = False
            comm_subexpr_elim = False
    
        if streaming:
            comm_subplan_elim = False
    
        ldf = self._ldf.optimization_toggle(
            type_coercion,
            predicate_pushdown,
            projection_pushdown,
            simplify_expression,
            slice_pushdown,
            comm_subplan_elim,
            comm_subexpr_elim,
            streaming,
            _eager,
        )
        if background:
            return InProcessQuery(ldf.collect_concurrently())
    
>       return wrap_df(ldf.collect())
E       polars.exceptions.ComputeError: datatypes of join keys don't match - `timestamp`: null on left does not match `timestamp`: i64 on right

/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/polars/lazyframe/frame.py:1749: ComputeError
----------------------------- Captured stdout call -----------------------------
Get historical data, across many exchanges & pairs: begin.
  Data start: timestamp=1687046400000, dt=2023-06-18_00:00:00.000
  Data fin: timestamp=1687132800000, dt=2023-06-19_00:00:00.000
  Update all rawohlcv files: begin
    Update rawohlcv file at exchange=binanceus, pair=BTC/USDT.
      filename=/tmp/pytest-of-runner/pytest-0/test_get_mergedohlcv_df_happyp0/binanceus_BTC-USDT_5m.parquet
      No file exists yet, so will fetch all data
      Aim to fetch data from start time: timestamp=1687046400000, dt=2023-06-18_00:00:00.000
      Fetch 1000 pts from timestamp=1687046400000, dt=2023-06-18_00:00:00.000
  Just saved df with 289 rows to new file /tmp/pytest-of-runner/pytest-0/test_get_mergedohlcv_df_happyp0/binanceus_BTC-USDT_5m.parquet
    Update rawohlcv file at exchange=binanceus, pair=ETH/USDT.
      filename=/tmp/pytest-of-runner/pytest-0/test_get_mergedohlcv_df_happyp0/binanceus_ETH-USDT_5m.parquet
      No file exists yet, so will fetch all data
      Aim to fetch data from start time: timestamp=1687046400000, dt=2023-06-18_00:00:00.000
      Fetch 1000 pts from timestamp=1687046400000, dt=2023-06-18_00:00:00.000
  Just saved df with 289 rows to new file /tmp/pytest-of-runner/pytest-0/test_get_mergedohlcv_df_happyp0/binanceus_ETH-USDT_5m.parquet
    Update rawohlcv file at exchange=kraken, pair=BTC/USDT.
      filename=/tmp/pytest-of-runner/pytest-0/test_get_mergedohlcv_df_happyp0/kraken_BTC-USDT_5m.parquet
      No file exists yet, so will fetch all data
      Aim to fetch data from start time: timestamp=1687046400000, dt=2023-06-18_00:00:00.000
      Fetch 1000 pts from timestamp=1687046400000, dt=2023-06-18_00:00:00.000
  Just saved df with 0 rows to new file /tmp/pytest-of-runner/pytest-0/test_get_mergedohlcv_df_happyp0/kraken_BTC-USDT_5m.parquet
  Update all rawohlcv files: done
  Load rawohlcv file.
  Merge rawohlcv dataframes.
=============================== warnings summary ===============================
...
=========================== short test summary info ============================
FAILED pdr_backend/lake/test/test_ohlcv_data_factory.py::test_get_mergedohlcv_df_happypath - polars.exceptions.ComputeError: datatypes of join keys don't match - `timestamp`: null on left does not match `timestamp`: i64 on right
======= 1 failed, [327](https://github.com/oceanprotocol/pdr-backend/actions/runs/7472573608/job/20335037499#step:8:328) passed, 3 skipped, 9 warnings in 101.22s (0:01:41) =======
Error: Process completed with exit code 1.
```

--- a/pdr_backend/lake/test/test_ohlcv_data_factory.py
+++ b/pdr_backend/lake/test/test_ohlcv_data_factory.py
@@ -3,9 +3,10 @@ import time
 from typing import List
 from unittest.mock import Mock, patch
 
+from enforce_typing import enforce_types
 import numpy as np
 import polars as pl
-from enforce_typing import enforce_types
+import pytest
 
 from pdr_backend.cli.arg_feed import ArgFeed
 from pdr_backend.lake.constants import TOHLCV_SCHEMA_PL
@@ -28,29 +29,18 @@ MS_PER_5M_EPOCH = 300000
 # ====================================================================
 # test update of rawohlcv files
 
-
-def test_update_rawohlcv_files1(tmpdir):
-    _test_update_rawohlcv_files("2023-01-01_0:00", "2023-01-01_0:00", tmpdir, n_uts=1)
-
-
-def test_update_rawohlcv_files2(tmpdir):
-    _test_update_rawohlcv_files("2023-01-01_0:00", "2023-01-01_0:05", tmpdir, n_uts=2)
-
-
-def test_update_rawohlcv_files3(tmpdir):
-    _test_update_rawohlcv_files("2023-01-01_0:00", "2023-01-01_0:10", tmpdir, n_uts=3)
-
-
-def test_update_rawohlcv_files4(tmpdir):
-    _test_update_rawohlcv_files("2023-01-01_0:00", "2023-01-01_0:45", tmpdir, n_uts=10)
-
-
-def test_update_rawohlcv_files5(tmpdir):
-    _test_update_rawohlcv_files("2023-01-01", "2023-06-21", tmpdir, n_uts=">1K")
-
-
 @enforce_types
-def _test_update_rawohlcv_files(st_timestr: str, fin_timestr: str, tmpdir, n_uts):
+@pytest.mark.parametrize(
+    "st_timestr, fin_timestr, n_uts",
+    [
+        ("2023-01-01_0:00", "2023-01-01_0:00", 1),
+        ("2023-01-01_0:00", "2023-01-01_0:05", 2),
+        ("2023-01-01_0:00", "2023-01-01_0:10", 3),
+        ("2023-01-01_0:00", "2023-01-01_0:45", 10),
+        ("2023-01-01", "2023-06-21", ">1K"),
+    ]
+)
+def test_update_rawohlcv_files(st_timestr: str, fin_timestr: str, n_uts, tmpdir):
     """
     @arguments
       n_uts -- expected # timestamps. Typically int. If '>1K', expect >1000
@@ -62,17 +52,13 @@ def _test_update_rawohlcv_files(st_timestr: str, fin_timestr: str, tmpdir, n_uts
         return since + i * MS_PER_5M_EPOCH
 
     def _uts_in_range(st_ut: int, fin_ut: int) -> List[int]:
+        return _uts_from_since(fin_ut=fin_ut, st_ut=st_ut, limit_N=100000)
+    
+    def _uts_from_since(fin_ut: int, st_ut: int, limit_N: int) -> List[int]:
         return [
             _calc_ut(st_ut, i)
-            for i in range(100000)  # assume <=100K epochs
-            if _calc_ut(st_ut, i) <= fin_ut
-        ]
-
-    def _uts_from_since(cur_ut: int, since_ut: int, limit_N: int) -> List[int]:
-        return [
-            _calc_ut(since_ut, i)
             for i in range(limit_N)
-            if _calc_ut(since_ut, i) <= cur_ut
+            if _calc_ut(st_ut, i) <= fin_ut
         ]
 
     # setup: exchange

--- a/pdr_backend/lake/test/test_ohlcv_data_factory.py
+++ b/pdr_backend/lake/test/test_ohlcv_data_factory.py
@@ -52,10 +52,7 @@ def test_update_rawohlcv_files(st_timestr: str, fin_timestr: str, n_uts, tmpdir)
         """Return a ut : unix time, in ms, in UTC time zone"""
         return since + i * MS_PER_5M_EPOCH
 
-    def _uts_in_range(st_ut: int, fin_ut: int) -> List[int]:
-        return _uts_from_since(fin_ut=fin_ut, st_ut=st_ut, limit_N=100000)
-
-    def _uts_from_since(fin_ut: int, st_ut: int, limit_N: int) -> List[int]:
+    def _uts_in_range(st_ut: int, fin_ut: int, limit_N=100000) -> List[int]:
         return [
             _calc_ut(st_ut, i) for i in range(limit_N) if _calc_ut(st_ut, i) <= fin_ut
         ]
@@ -67,7 +64,7 @@ def test_update_rawohlcv_files(st_timestr: str, fin_timestr: str, n_uts, tmpdir)
 
         # pylint: disable=unused-argument
         def fetch_ohlcv(self, since, limit, *args, **kwargs) -> list:
-            uts: List[int] = _uts_from_since(self.cur_ut, since, limit)
+            uts: List[int] = _uts_in_range(since, self.cur_ut, limit)
             return [[ut] + [1.0] * 5 for ut in uts]  # 1.0 for open, high, ..
 
     ss, factory = _lake_ss_1feed(

--- a/pdr_backend/lake/test/test_ohlcv_data_factory.py
+++ b/pdr_backend/lake/test/test_ohlcv_data_factory.py
@@ -29,6 +29,7 @@ MS_PER_5M_EPOCH = 300000
 # ====================================================================
 # test update of rawohlcv files
 
+
 @enforce_types
 @pytest.mark.parametrize(
     "st_timestr, fin_timestr, n_uts",
@@ -38,7 +39,7 @@ MS_PER_5M_EPOCH = 300000
         ("2023-01-01_0:00", "2023-01-01_0:10", 3),
         ("2023-01-01_0:00", "2023-01-01_0:45", 10),
         ("2023-01-01", "2023-06-21", ">1K"),
-    ]
+    ],
 )
 def test_update_rawohlcv_files(st_timestr: str, fin_timestr: str, n_uts, tmpdir):
     """
@@ -53,12 +54,10 @@ def test_update_rawohlcv_files(st_timestr: str, fin_timestr: str, n_uts, tmpdir)
 
     def _uts_in_range(st_ut: int, fin_ut: int) -> List[int]:
         return _uts_from_since(fin_ut=fin_ut, st_ut=st_ut, limit_N=100000)
-    
+
     def _uts_from_since(fin_ut: int, st_ut: int, limit_N: int) -> List[int]:
         return [
-            _calc_ut(st_ut, i)
-            for i in range(limit_N)
-            if _calc_ut(st_ut, i) <= fin_ut
+            _calc_ut(st_ut, i) for i in range(limit_N) if _calc_ut(st_ut, i) <= fin_ut
         ]
 
     # setup: exchange

